virttest/compat.py

```diff
@@ -1,28 +1,83 @@
 import argparse
 
+from avocado.core.settings import settings
 
-def get_opt(opt, name):
-    """
-    Compatibility handler for options in either argparse.Namespace or dict
 
-    :param opt: either an argpase.Namespace instance or a dict
-    :param name: the name of the attribute or key
-    """
-    if isinstance(opt, argparse.Namespace):
-        return getattr(opt, name, None)
-    else:
-        return opt.get(name)
+def is_registering_settings_required():
+    """Checks the characteristics of the Avocado settings API.
 
+    And signals if the explicit registration of options is required, along
+    with other API details that should be followed.
 
-def set_opt(opt, name, value):
-    """
-    Compatibility handler for options in either argparse.Namespace or dict
+    The heuristic used here is to check for methods that are only present
+    in the new API, and should be "safe enough".
 
-    :param opt: either an argpase.Namespace instance or a dict
-    :param name: the name of the attribute or key
-    :param value: the value to be set
+    TODO: remove this once support for Avocado releases before 81.0,
+    including 69.x LTS is dropped.
     """
-    if isinstance(opt, argparse.Namespace):
-        setattr(opt, name, value)
-    else:
+    return (hasattr(settings, 'add_argparser_to_option') and
+            hasattr(settings, 'register_option') and
+            hasattr(settings, 'as_json'))
+
+
+if is_registering_settings_required():
+    def get_opt(opt, name):
+        """
+        Compatibility handler to Avocado with configuration as dict
+
+        :param opt: a configuration dict, usually from settings.as_dict()
+        :param name: the name of the configuration key, AKA namespace
+        """
+        return opt.get(name)
+
+    def set_opt(opt, name, value):
+        """
+        Compatibility handler to Avocado with configuration as dict
+
+        :param opt: a configuration dict, usually from settings.as_dict()
+        :param name: the name of the configuration key, AKA namespace
+        :param value: the value to be set
+        """
         opt[name] = value
+
+    def set_opt_from_settings(opt, section, key, **kwargs):
+        """No-op, default values are set at settings.register_option()."""
+        pass
```

Comment:
What is meant to be achieved with this noop function? If the users call it should they be warned that they are doing it in the wrong way and they should simply register the options?

Too bad that there are now three different versions of settings within Avocado VT so the logic becomes so complicated and requires much extra handling. Hopefully we can drop some of these cases in the future.

Reply:
> What is meant to be achieved with this noop function? If the users call it should they be warned that they are doing it in the wrong way and they should simply register the options?
> 

When Avocado-VT is not using the new `avocado.core.future.settings`, it sets the default value from the configuration file.  This is "opposing" implementation that does that:

```python
    def set_opt_from_settings(opt, section, key, **kwargs):
        """Sets option default value from the configuration file."""
        value = settings.get_value(section, key, **kwargs)
        namespace = '%s.%s' % (section, key)
        set_opt(opt, namespace, value)

```

But, when using `avocado.core.future.settings`, the settings and their default values are kept within the `Settings` instance itself at registration time, so this is not necessary and becomes an empty place holder.  I guess other approach could have used, but this seemed to be the best way: have the same code elsewhere and determine the behavior at a central place.    

> Too bad that there are now three different versions of settings within Avocado VT so the logic becomes so complicated and requires much extra handling. Hopefully we can drop some of these cases in the future.

I absolutely agree.  Stay tuned, because we do have plans for the continuing development of Avocado-VT, that we'll propose after the LTS release. Hopefully it will make things easier for all of us in the future.

Reply:
> But, when using `avocado.core.future.settings`, the settings and their default values are kept within the `Settings` instance itself at registration time, so this is not necessary and becomes an empty place holder. I guess other approach could have used, but this seemed to be the best way: have the same code elsewhere and determine the behavior at a central place.

I see, the good thing is that this entire annihilation magic behavior will get simplified in the future.

> > Too bad that there are now three different versions of settings within Avocado VT so the logic becomes so complicated and requires much extra handling. Hopefully we can drop some of these cases in the future.
> 
> I absolutely agree. Stay tuned, because we do have plans for the continuing development of Avocado-VT, that we'll propose after the LTS release. Hopefully it will make things easier for all of us in the future.

In any case on our side we will do whatever we can to drop any dependencies we can and get as close to vanilla avocado as possible. But this is a matter of future discussions and in any case I will be curious to hear what you come up with.